pkg/hoist/artifact_cache.go

```diff
@@ -0,0 +1,117 @@
+package hoist
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/uri"
+	"github.com/square/p2/pkg/util"
+)
+
+// ArtifactCacher keeps a directory of Hoist artifacts that can be reused for future downloads.
+// This allows artifacts to be cheaply shared amongst different pods, especially in cases where
+// the artifacts contain base libraries and runtimes. For example, a pod may include a Java 8
+// artifact that is reused across other pods.
+//
+// The cache assumes that Hoist artifact URIs are unique at the basename, ie. comply with the
+// convention of {artifact_name}_{artifact_SHA}.tar.gz pattern. No attempt is made to distinguish
+// artifacts that are named the same but contain different content. (Checksumming is presumed to
+// be done by the Launchable itself)
+//
+// Once a cached copy of an artifact is created, it will be hardlinked into the final destination
+// URI provided by the caller. This allows the ArtifactCacher to safely prune without worrying about
+// whether the artifact is still in use by a particular Launchable (say, as the Last link). However,
+// as a result of this strategy, pruning the cache dir will not be sufficient to free system disk
+// resources as other links to the same tars will also need to be removed.
+type ArtifactCacher struct {
+	cacheDir      string
+	nestedFetcher Fetcher
+	Logger        *logging.Logger
+}
+
+// Complies with the Fetcher interface
+func (a *ArtifactCacher) Fetch(fromURI, toURI string) error {
+	base := filepath.Base(toURI)
```

Comment:
`toURI` should probably be `toPath`? That would match the signature of the existing uri.URICopy.